include/watchdog.h

```diff
@@ -1,116 +1,168 @@
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
+ * num_of_channels - number of channels that should be enabled on
+ *		     given watchdog instance.
  */
 struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+	u32_t *timeouts;
+	void (*interrupt_fn)(struct device *dev, u8_t channel);
+	u8_t operation;
+	u8_t num_of_channels;
 };
 
-typedef void (*wdt_api_enable)(struct device *dev);
-typedef void (*wdt_api_disable)(struct device *dev);
+typedef int (*wdt_api_enable)(struct device *dev);
+typedef int (*wdt_api_disable)(struct device *dev);
 typedef int (*wdt_api_set_config)(struct device *dev,
 				  struct wdt_config *config);
 typedef void (*wdt_api_get_config)(struct device *dev,
 				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
+typedef void (*wdt_api_reload)(struct device *dev, u8_t channel);
 
 struct wdt_driver_api {
 	wdt_api_enable enable;
 	wdt_api_disable disable;
-	wdt_api_get_config get_config;
 	wdt_api_set_config set_config;
+	wdt_api_get_config get_config;
 	wdt_api_reload reload;
 };
 
-static inline void wdt_enable(struct device *dev)
+/*
+ * @brief Enable watchdog instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance has not been configured yet.
+ * @retval -EBUSY if watchdog instance has been already enabled.
+ */
+static inline int wdt_enable(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->enable(dev);
+	return api->enable(dev);
 }
 
-static inline void wdt_disable(struct device *dev)
+/*
+ * @brief Disable watchdog instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance is not enabled.
+ * @retval -EPERM if watchdog can not be disabled directly by application code.
+ */
+static inline int wdt_disable(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->disable(dev);
+	return api->disable(dev);
 }
 
-static inline void wdt_get_config(struct device *dev,
-				  struct wdt_config *config)
+/*
+ * @brief Configure watchdog instance.
+ *
+ * @retval 0 if successful.
+ * @retval -ENOTSUP if any of the set operation settings is not supported.
+ * @retval -E2BIG if number of channels is too big.
+ * @retval -EINVAL if any of the timeout values is out of possible range.
+ * @retval -ENOEXEC if WDT_OP_MODE_INTERRUPT_RESET is set but interrupt function
+ *		    pointer is NULL.
+ */
+static inline int wdt_set_config(struct device *dev,
+				 struct wdt_config *config)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->get_config(dev, config);
+	return api->set_config(dev, config);
 }
 
-static inline int wdt_set_config(struct device *dev,
-				 struct wdt_config *config)
+/**
+ * @brief Get configuration of watchdog.
+ */
+static inline void wdt_get_config(struct device *dev,
```

Comment:
I wonder if we need this function at all. It forces the device driver to store internally the whole struct wdt_config. Also it is not clear what should be done with config.timeouts array. In struct wdt_config we store only the pointer. It does not have to point to a static variable.

Reply:
Indeed, not sure this one is useful, it comes from legacy API after all. What would be the use case for this function?